docs/source/usecases/FunctionPointers.md

```diff
@@ -0,0 +1,286 @@
+
+```c++
+#include <Kokkos_Core.hpp>
+
+KOKKOS_INLINE_FUNCTION void foo() {
+  KOKKOS_IF_ON_HOST(printf("foo called from host\n");)
+  KOKKOS_IF_ON_DEVICE(printf("foo called from device\n");)
+}
+
+struct SomeClass {
+  void (*bar)();
+  KOKKOS_FUNCTION void print() const {
+    bar();
+  }
+};
+
+int main(int argc, char* argv[]) {
+  Kokkos::initialize(argc, argv);
+  {
+    SomeClass A;
+    A.bar = &foo;
+    // Call it plain on host
+    A.print();
+
+    // Call it inside a host parallel for
+    printf("I can use the function pointer in a host parallel_for!\n");
+    Kokkos::parallel_for(Kokkos::RangePolicy<Kokkos::DefaultHostExecutionSpace>(0,1), KOKKOS_LAMBDA(int) {
+      A.print();
+    });
+    Kokkos::fence();
+    printf("Now I will crash if we compiled for CUDA/HIP\n");
+    // Try to call it on device
+    Kokkos::parallel_for(Kokkos::RangePolicy<Kokkos::DefaultExecutionSpace>(0,1), KOKKOS_LAMBDA(int) {
+      A.print();
+    });
+    Kokkos::fence();
+    printf("Never got here in CUDA/HIP\n");
+
+  }
+  Kokkos::finalize();
+}
+```
+
+This worked on host (both inside and outside a `parallel_for` but crashes inside a device kernel.
+Here is the output:
+
+```
+foo called from host
+I can use the function pointer in a host parallel_for!
+foo called from host
+Now I will crash if we compiled for CUDA/HIP
+cudaDeviceSynchronize() error( cudaErrorIllegalAddress): an illegal memory access was encountered /home/crtrott/Kokkos/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:153
+Backtrace:
+[0x4422e3] 
+[0x4249d5] 
+[0x43085d] 
+[0x430a5e] 
+[0x429855] 
+[0x40843e] 
+[0x7f7fca5ba7e5] __libc_start_main
+[0x409cce] 
+Aborted (core dumped)
+```
+
+*The function pointer we created was for a host function: you can not use it on device same as you can't dereference data pointers to host data!*
+
+## Getting a device function pointer
+
+We actually can get the pointer to the device version of our function `foo`. But we can only do that *on the device*!
+That means we need to run a little Kokkos kernel get the function pointer there, and copy it somehow back to the host so we can set the host object that way.
+
+To do this we need a device `View` of a `SomeClass` instance, set the function pointer in the device code, and `deep_copy` it back into our host instance.
+
+```c++
+    SomeClass A;
+    Kokkos::View<SomeClass> A_v("A_v");
+    // Now init the function pointer on device:
+    Kokkos::parallel_for(Kokkos::RangePolicy<Kokkos::DefaultExecutionSpace>(0,1), KOKKOS_LAMBDA(int) {
+      A_v().bar = &foo;
+    });
+    // copy the device instance back to the host
+    Kokkos::deep_copy(A, A_v);
+```
+
+We are leveraging here the fact that `deep_copy` allows you to copy from and to a host scalar value, from a `View` of rank-0.
+
+If we do this `A` will contain a function pointer to a device function. That means we can capture `A` into a parallel region and execute it on the device, but now it will crash on the host.
```

Comment:
```suggestion
If we do this, `A` will contain a function pointer to a device function. That means we can capture `A` into a parallel region and execute it on the device, but now it will crash on the host.
```